samza-core/src/main/java/org/apache/samza/storage/blobstore/BlobStoreBackupManager.java

```diff
@@ -0,0 +1,349 @@
+            prevStoreSnapshotIndexesFuture.get(0, TimeUnit.MILLISECONDS);
+
+        // get the previous store directory contents
+        DirIndex prevDirIndex;
+
+        if (prevStoreSnapshotIndexes.containsKey(storeName)) {
+          prevDirIndex = prevStoreSnapshotIndexes.get(storeName).getRight().getDirIndex();
+        } else {
+          // no previous SnapshotIndex means that this is the first commit for this store. Create an empty DirIndex.
+          prevDirIndex = new DirIndex(checkpointDir.getName(), Collections.emptyList(), Collections.emptyList(),
+              Collections.emptyList(), Collections.emptyList());
+        }
+
+        long dirDiffStartTime = System.nanoTime();
+        // get the diff between previous and current store directories
+        DirDiff dirDiff = DirDiffUtil.getDirDiff(checkpointDir, prevDirIndex, DirDiffUtil.areSameFile(false));
+        metrics.storeDirDiffNs.get(storeName).update(System.nanoTime() - dirDiffStartTime);
+
+        DirDiff.Stats stats = DirDiff.getStats(dirDiff);
+        updateStoreDiffMetrics(storeName, stats);
+        metrics.filesToUpload.getValue().addAndGet(stats.filesAdded);
+        metrics.bytesToUpload.getValue().addAndGet(stats.bytesAdded);
+        // Note: FilesRemaining metric is set to FilesAdded in the beginning of the current upload and then counted down
+        // for each upload.
+        metrics.filesRemaining.getValue().addAndGet(stats.filesAdded);
+        metrics.bytesRemaining.getValue().addAndGet(stats.bytesAdded);
+        metrics.filesToRetain.getValue().addAndGet(stats.filesRetained);
+        metrics.bytesToRetain.getValue().addAndGet(stats.bytesRetained);
+
+        // upload the diff to the blob store and get the new directory index
+        CompletionStage<DirIndex> dirIndexFuture = blobStoreUtil.putDir(dirDiff, snapshotMetadata);
+
+        CompletionStage<SnapshotIndex> snapshotIndexFuture =
+            dirIndexFuture.thenApplyAsync(dirIndex -> {
+              LOG.trace("Dir upload complete. Returning new SnapshotIndex for task: {} store: {}.", taskName, storeName);
+              Optional<String> prevSnapshotIndexBlobId =
+                  Optional.ofNullable(prevStoreSnapshotIndexes.get(storeName)).map(Pair::getLeft);
+              return new SnapshotIndex(clock.currentTimeMillis(), snapshotMetadata, dirIndex, prevSnapshotIndexBlobId);
+            }, executor);
+
+        // upload the new snapshot index to the blob store and get its blob id
+        CompletionStage<String> snapshotIndexBlobIdFuture =
+            snapshotIndexFuture
+                .thenComposeAsync(si -> {
+                  LOG.trace("Uploading Snapshot index for task: {} store: {}", taskName, storeName);
+                  return blobStoreUtil.putSnapshotIndex(si);
+                }, executor);
+
+        // update the map of storeName to previous snapshot index storeToSCMAndSnapshotIndexPairFutures which is temporary
```

Comment:
Run on sentence is hard to parse, can you simplify? Don't need to mention variable names in comment, it's obvious from context. Just say what is happening here and why. 

There is actually no update happening here at all, is this comment misplaced (should be on 244)?

Maybe we can delete the comment entirely?